src/forms/Typeahead.jsx

```diff
@@ -14,32 +14,91 @@ const defaultItemToString = itemValue => (typeof itemValue === 'string' ? itemVa
  * @module Typeahead
  */
 class Typeahead extends React.PureComponent {
+	constructor(props) {
+		super(props);
+
+		this.handleSelection = this.handleSelection.bind(this);
+		// this.handleSingleSelection = this.handleSingleSelection.bind(this);
+		// this.handleMultiselectSelect = this.handleMultiselectSelect.bind(this);
+		this.stateReducer = this.stateReducer.bind(this);
+	}
+
+	// handleMultiselectSelect(item) {
+	// 	this.props.onSelect(item, [...this.props.multiSelectValues, item]);
+	// }
+
+	// handleSingleSelection(selectedItem, stateAndHelpers) {
+	// 	!this.props.multiSelect && this.props.onSelect && this.props.onSelect(selectedItem, stateAndHelpers);
+	// }
+
+	handleSelection(selectedItem, stateAndHelpers) {
+		if(this.props.multiSelect) {
+			this.props.onSelect(selectedItem, [...this.props.multiSelectValues, selectedItem]);
+		} else {
+			this.props.onSelect && this.props.onSelect(selectedItem, stateAndHelpers);
+		}
+	}
+
+	stateReducer(state, changes) {
+		switch (changes.type) {
+			case Downshift.stateChangeTypes.keyDownEnter:
+			case Downshift.stateChangeTypes.clickItem:
+				this.handleSelection(changes.selectedItem);
+
+				return {
+					...changes,
+					selectedItem: this.props.multiSelectValues ? this.props.multiSelectValues : changes.selectedItem,
+					isOpen: this.props.openOnSelect,
+				};
+
+			case Downshift.stateChangeTypes.blurInput:
+				return {
+					...changes,
+					isOpen: this.props.openOnSelect,
+				};
+
+			default:
+				return changes;
+		}
+	}
 
 	render() {
 		const {
 			inputProps,
 			height,
 			items,
+			multiSelect,
+			multiSelectValues,
+			openOnFocus,
+			openOnSelect,
 			itemToString,
 			...other
 		} = this.props;
 
+		delete other.onSelect;
```

Comment:
Since you're already destructuring the props above, you could do the following...
```js
const { 
   // ... all the other props before this
   itemToString,
   onSelect,
   ...other
} = this.props;
```
to remove onSelect from other.